notes/algorithms.md

```diff
@@ -0,0 +1,100 @@
+#Listy
+
+- węzły i krawędzie
+- określona maksymalna liczba wychodzących strzałek
+-
+
+###SingleLinkedList [SLL]
+
+- węzeł jest klasą, która trzyma dane
+- połączone pojedynczymi strzałkami
+
+- potrzebna jest klasa sterująca z metodami - SingleLinkedList
+- klasa nadzorcza łączy się z danymi przez uchwyt - handle
+
+Dygresja - strażnik sentinel - w javie null pointer jest strażnikiem
+
+- nie zabijaj sentinela - nie przekształcać danych na sentinela - np przy decrement.
+
+
+    lista jest pewną formą grafu - różnią się tym od grafów tym, że w listach istnieje porządek
+    
+DoubleLinkedList - jest wtedy jak lista zawiera 2 przeciwne porządki - do przodu i do tyłu
+
+referencji może być wiele zmieniając porządek sortowania listy - kolejność dodania (techniczna) 
+
+SLL - 1 krawędź
+    - łatwa
+    - mało 
+
+    - niecofalna
+    
+DLL - head, tail??
+    - chodzi w dwóch kierunkach
+    - można zacząć gdziekolwiek   
+ 
+ 
+ Metryki kodu:
+ - liczba linijek kodu
+ - liczba `if` - z niej wynika liczba potrzebych unitestów
```

Comment:
Code complexity: liczba rozgałęzień kodu: `if', `for`, `&&`. Nie więcej niż 20 w metodzie. W zasadzie to 10